File: pset2/vigenere/vigenere.c
#include <cs50.h>
#include <stdio.h>
#include <string.h>
#include <ctype.h>

// select key from a string
int selectK(string k , int index)
{
    if(isupper(k[index]))
    {
        return k[index]-65;
    }
    else
    {
        return k[index]-97;
    }
}

int main(int argc, string argv[])
{
    // ensure proper usage
    if(argc < 2 || argc > 2)
    {
        printf("Usage: ./vigenere k\n");
        exit(1);
    }
    else
    {
        string keyword = argv[1];
        // check if all strings are alphabetical
        for(int i = 0 , n = strlen(keyword) ; i < n ; i ++)
        {
            if(!isalpha(keyword[i]))
            {
                printf("Illegal input\n");
                exit(1);
            }
        }
        // variable for length of keyword;
        int lengthK = strlen(keyword);
        // variable for k
        int k;
        // variable for chosing K
        int indexK = 0;
        // variable for alphabetical 
        int index;
        // variable for saving
        int temp;
        printf("plaintext: ");
        string s = get_string();
        for(int i = 0 , n = strlen(s) ; i < n ; i ++)
        {
            if(indexK==lengthK)
            indexK = 0;
            if(isalpha(s[i]))
            {
                //A letter's part
                if(isupper(s[i]))
                {
                    index = s[i]-65;
                    k = selectK(keyword,indexK);
                    temp = ( index + k ) % 26;
                    s[i] = (char)(temp+65);
                }
                //a letter's part
                else
                {
                    index = s[i]-97;
                    k = selectK(keyword,indexK);
                    temp = ( index + k ) % 26;
                    s[i] = (char)(temp+97);
                }
            indexK++;
            }
        }
        printf("ciphertext: %s\n", s);
        return 0;
    }
}

File: pset6/vigenere.py
import cs50
import sys

# select key from a string
def selectK(k, index):
    if str.isupper(k[index]):
        return ord(k[index]) - 65
    else:
        return ord(k[index]) - 97

def main():
    # ensure proper usage
    if len(sys.argv) != 2:
        sys.exit("Usage: ./vigenere k")
    # ensure proper usage
    keyword = sys.argv[1]
    for i in keyword:
        if not str.isalpha(i):
            print("k must be letters")
            sys.exit(1)
    
    k_length = len(keyword)
    k = 0
    k_index = 0
    index = 0
    temp = 0
    ans = ""
    
    print("plaintext:  ", end = "")
    s = cs50.get_string()
    for i in s:
        if k_index == k_length:
            k_index = 0
        # check if it's letter
        if str.isalpha(i):
            # if it's uppercase
            if str.isupper(i):
                index = ord(i) - 65
                k = selectK(keyword, k_index)
                temp = ( index + k ) % 26
                ans += chr(temp + 65)
            # if it's lowercase
            else:
                index = ord(i) - 97
                k = selectK(keyword, k_index)
                temp = ( index + k ) % 26
                ans += chr(temp + 97)
            k_index += 1
        else:
            ans += i
    
    print("ciphertext: {}".format(ans))
    return

if __name__ == '__main__':
    main()